hal/api/Emac.h

```diff
@@ -0,0 +1,173 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_EMAC_H
+#define MBED_EMAC_H
+
+#include "platform.h"
+
+#if DEVICE_EMAC
+
+#include "stack_mem.h"
+
+typedef struct emac_interface emac_interface_t;
+
+/**
+ * EmacInterface
+ *
+ * This interface should be used to abstract low level access to networking hardware
+ */
+
+/**
+ * Network interface capabilities
+ *
+ * Right now the flags are LWIP specific, but in the future we may need to support more stacks.
+ */
+#define EMAC_CAP_BROADCAST    0x01U     /*!< Interface has broadcast capabilities */
+#define EMAC_CAP_ETHARP       0x02U     /*!< Interface supports ARP */
+#define EMAC_CAP_ETHERNET     0x04U     /*!< Ethernet device */
+#define EMAC_CAP_IGMP         0x08U     /*!< Interface has IGMP capabilities */
```

Comment:
It could be extended in future, I agree. But I'm wary of being over-ambitious. Especially as I don't see the use cases. For example "PPP" pre-supposes an external PPP implementation to the stack. Does that seem likely? And it seems unlikely we'd ever use this for 802.15.4 (6LoWPAN).

You could certainly reserve some type fields ready for expansion - MAC + PHY type, and flags? At the minute the MAC type is "Ethernet/802.3" for both Wi-Fi and Ethernet (because you're expecting 802.11-format frames, not 802.11 format frames). A radically different MAC like PPP or 802.15.4 could be indicated there. The PHY type could may indicate the Ethernet/Wi-Fi distinction (together with the actual speed negotiated).